Add model flexibility paragraph and minor improvements (#3435)
huggingface/blog: cross-origin-storage.md

console.log(result);
```

Note the `experimental_` prefix on the flag. It's intentional and signals that the underlying browser API has not yet been standardized and may change without a major version bump. With that flag set, Transformers.js resolves the SHA-256 hash for each [Xet-tracked](https://huggingface.co/docs/hub/en/xet/index) model file (the large ONNX weight files) by fetching the raw Xet pointer ([example raw pointer file](https://huggingface.co/Xenova/whisper-tiny.en/raw/main/onnx/decoder_model.onnx)) and extracting its `oid sha256:` field. It then uses that hash as the key for `navigator.crossOriginStorage`. If the model is already in COS (because another site stored it there first), it's served instantly without a network round-trip. If not, it falls back to a regular download and stores the result in COS for the next caller. With the toy example, the advantage in practice is that `Xenova/whisper-tiny.en` and `Xenova/distilbert-base-uncased-finetuned-sst-2-english` (and of course `ort-wasm-simd-threaded.asyncify.wasm`) only ever need to cross the ether once, regardless of how many different origins ask for them.

### Model flexibility

> The toy example works just fine with `Xenova/whisper-tiny.en`, but of course you possibly wouldn't say no if the user already has [any of the other Whisper variants](https://huggingface.co/Xenova/models?search=whisper) in their COS cache. For example, the user might already have [`Xenova/whisper-large-v3`](https://huggingface.co/Xenova/whisper-large-v3), which, as the name suggests, is a lot larger than the tiny variant. Transformers.js's [Model Registry](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry) makes being flexible about your models possible. If you know your app's needs can be addressed by, for example, any of `Xenova/whisper-tiny.en`, `whisper-medium.en`, or `Xenova/whisper-large-v3`, you can check the registry for the associated files for each model, probe for their existence in the COS cache (which may partially or completely contain the model resources you need), and then take a decision what model to choose eventually. As COS support matures in Transformers.js, I expect there to develop ergonomic improvements for the model flexibility use case, like a direct COS integration for the [`ModelRegistry.is_cached()`](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry#modelregistryiscachedmodelid-options--promise--boolean-) API.

Author: ```suggestion
The toy example works just fine with `Xenova/whisper-tiny.en`, but of course you possibly wouldn't say no if the user already has [any of the other Whisper variants](https://huggingface.co/Xenova/models?search=whisper) in their COS cache. For example, the user might already have [`Xenova/whisper-large-v3`](https://huggingface.co/Xenova/whisper-large-v3), which, as the name suggests, is a lot larger than the tiny variant. Transformers.js's [Model Registry](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry) makes being flexible about your models possible. If you know your app's needs can be addressed by, for example, any of `Xenova/whisper-tiny.en`, `whisper-medium.en`, or `Xenova/whisper-large-v3`, you can check the registry for the associated files for each model, probe for their existence in the COS cache (which may partially or completely contain the model resources you need), and then take a decision what model to choose eventually. The [`ModelRegistry.is_cached()`](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry#modelregistryiscachedmodelid-options--promise--boolean-) API directly integrates with COS (and of course the Cache API), so this operation is really ergonomic.
```

console.log(result);
```

Note the `experimental_` prefix on the flag. It's intentional and signals that the underlying browser API has not yet been standardized and may change without a major version bump. With that flag set, Transformers.js resolves the SHA-256 hash for each [Xet-tracked](https://huggingface.co/docs/hub/en/xet/index) model file (the large ONNX weight files) by fetching the raw Xet pointer ([example raw pointer file](https://huggingface.co/Xenova/whisper-tiny.en/raw/main/onnx/decoder_model.onnx)) and extracting its `oid sha256:` field. It then uses that hash as the key for `navigator.crossOriginStorage`. If the model is already in COS (because another site stored it there first), it's served instantly without a network round-trip. If not, it falls back to a regular download and stores the result in COS for the next caller. With the toy example, the advantage in practice is that `Xenova/whisper-tiny.en` and `Xenova/distilbert-base-uncased-finetuned-sst-2-english` (and of course `ort-wasm-simd-threaded.asyncify.wasm`) only ever need to cross the ether once, regardless of how many different origins ask for them.

### Model flexibility

> The toy example works just fine with `Xenova/whisper-tiny.en`, but of course you possibly wouldn't say no if the user already has [any of the other Whisper variants](https://huggingface.co/Xenova/models?search=whisper) in their COS cache. For example, the user might already have [`Xenova/whisper-large-v3`](https://huggingface.co/Xenova/whisper-large-v3), which, as the name suggests, is a lot larger than the tiny variant. Transformers.js's [Model Registry](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry) makes being flexible about your models possible. If you know your app's needs can be addressed by, for example, any of `Xenova/whisper-tiny.en`, `whisper-medium.en`, or `Xenova/whisper-large-v3`, you can check the registry for the associated files for each model, probe for their existence in the COS cache (which may partially or completely contain the model resources you need), and then take a decision what model to choose eventually. The [`ModelRegistry.is_cached()`](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry#modelregistryiscachedmodelid-options--promise--boolean-) API directly integrates with COS (and of course the Cache API), so this operation is really ergonomic.

Reviewer: Ahh, clever!

Author: Hopefully this will become a pattern apps can rely on more in the future.

> The toy example works just fine with `Xenova/whisper-tiny.en`, but of course you possibly wouldn't say no if the user already has [any of the other Whisper variants](https://huggingface.co/Xenova/models?search=whisper) in their COS cache. For example, the user might already have [`Xenova/whisper-large-v3`](https://huggingface.co/Xenova/whisper-large-v3), which, as the name suggests, is a lot larger than the tiny variant. Transformers.js's [Model Registry](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry) makes being flexible about your models possible. If you know your app's needs can be addressed by, for example, any of `Xenova/whisper-tiny.en`, `whisper-medium.en`, or `Xenova/whisper-large-v3`, you can check the registry for the associated files for each model, probe for their existence in the COS cache (which may partially or completely contain the model resources you need), and then take a decision what model to choose eventually. The [`ModelRegistry.is_cached()`](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry#modelregistryiscachedmodelid-options--promise--boolean-) API directly integrates with COS (and of course the Cache API), so this operation is really ergonomic.

Reviewer 2: I would suggest the `is_pipeline_cached`: https://huggingface.co/docs/transformers.js/en/api/utils/model_registry#modelregistryispipelinecachedtask-modelid-options--promise--boolean-

When we know the expected pipeline we can be more efficient in foguring out what files need to be checked.

Author: Changed to the suggested function. PTAL.

> The toy example works just fine with `Xenova/whisper-tiny.en`, but of course you possibly wouldn't say no if the user already has [any of the other Whisper variants](https://huggingface.co/Xenova/models?search=whisper) in their COS cache. For example, the user might already have [`Xenova/whisper-large-v3`](https://huggingface.co/Xenova/whisper-large-v3), which, as the name suggests, is a lot larger than the tiny variant. Transformers.js's [Model Registry](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry) makes being flexible about your models possible. If you know your app's needs can be addressed by, for example, any of `Xenova/whisper-tiny.en`, `whisper-medium.en`, or `Xenova/whisper-large-v3`, you can check the registry for the associated files for each model, probe for their existence in the COS cache (which may partially or completely contain the model resources you need), and then take a decision what model to choose eventually. The [`ModelRegistry.is_cached()`](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry#modelregistryiscachedmodelid-options--promise--boolean-) API directly integrates with COS (and of course the Cache API), so this operation is really ergonomic.

Author: ```suggestion
The toy example works just fine with `Xenova/whisper-tiny.en`, but of course you possibly wouldn't say no if the user already has [any of the other Whisper variants](https://huggingface.co/Xenova/models?search=whisper) in their COS cache. For example, the user might already have [`Xenova/whisper-large-v3`](https://huggingface.co/Xenova/whisper-large-v3), which, as the name suggests, is a lot larger than the tiny variant. Transformers.js's [Model Registry](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry) makes being flexible about your models possible. If you know your app's needs can be addressed by, for example, any of `Xenova/whisper-tiny.en`, `whisper-medium.en`, or `Xenova/whisper-large-v3`, you can check the registry for the associated files for each model, probe for their existence in the COS cache (which may partially or completely contain the model resources you need), and then take a decision what model to choose eventually. The [`ModelRegistry.is_pipeline_cached()`](https://huggingface.co/docs/transformers.js/en/api/utils/model_registry#modelregistryispipelinecachedtask-modelid-options--promise--boolean-) API directly integrates with COS (and of course the Cache API), so this operation is really ergonomic.
```
